node/src/utils/round_robin.rs

```diff
@@ -115,70 +113,31 @@ struct Slot<K> {
     tickets: usize,
 }
 
-impl<I, K> WeightedRoundRobin<I, K>
+#[derive(Debug, Serialize)]
+/// A dump of the internal queues.
+pub struct QueueDump<'a, K, I>
 where
-    I: Serialize,
-    K: Copy + Clone + Eq + Hash + IntoEnumIterator + Serialize,
+    K: Hash + Eq,
 {
-    /// Create a snapshot of the queue by locking it and serializing it.
-    ///
-    /// The serialized events are streamed directly into `serializer`.
-    ///
-    /// # Warning
-    ///
-    /// This function locks all queues in the order defined by the order defined by
-    /// `IntoEnumIterator`. Calling it multiple times in parallel is safe, but other code that locks
-    /// more than one queue at the same time needs to be aware of this.
-    pub async fn snapshot<S: Serializer>(&self, serializer: S) -> Result<(), S::Error> {
-        // Lock all queues in order get a snapshot, but release eagerly. This way we are guaranteed
-        // to have a consistent result, but we also allow for queues to be used again earlier.
-        let mut locks = Vec::new();
-
-        for kind in K::into_enum_iter() {
-            let queue_guard = self
-                .queues
-                .get(&kind)
-                .expect("missing queue while snapshotting")
-                .queue
-                .lock()
-                .await;
-
-            locks.push((kind, queue_guard));
-        }
-
-        let mut map = serializer.serialize_map(Some(locks.len()))?;
-
-        // By iterating over the guards, they are dropped in order.
-        for (kind, guard) in locks {
-            let vd = &*guard;
-            map.serialize_key(&kind)?;
-            map.serialize_value(vd)?;
-        }
-        map.end()?;
-
-        Ok(())
-    }
+    queues: HashMap<K, &'a VecDeque<I>>,
 }
 
 impl<I, K> WeightedRoundRobin<I, K>
 where
     I: Debug,
     K: Copy + Clone + Eq + Hash + IntoEnumIterator + Debug,
 {
-    /// Dump the contents of the queues (`Debug` representation) to a given file.
-    pub async fn debug_dump(&self, file: &mut File) -> Result<(), io::Error> {
+    /// Dump the queue contents to the given dumper function.
+    pub async fn dump<F: FnOnce(&QueueDump<K, I>)>(&self, dumper: F) {
```

Comment:
So that is a shorthand for a function that returns `Unit`? Didn't know that.

Reply:
Yes, the `-> ()` is optional. I'd wager you're writing the shorthand every time you define a function without a return type already :)